feat(syntax_operations): add crate `oxc_syntax_operations`

Boshen opened the pull request:
I intend to add constant folding and eval logic to this crate. There are downstream tools that require these functionalities alone.

It's also reasonable to move these traits out of the `ast` crate.

Boshen commented:
* **#6202** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6202?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Boshen and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Oct 5, 2:51 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/6202) disabled "merge when ready" on this PR due to: a merge conflict with the target branch; resolve the conflict and try again..

overlookmotel commented:
This looks like the relevant part of Miri's error message:

> /home/runner/work/oxc/oxc/crates/oxc_transformer/src/plugins/replace_global_defines.rs:209:48
> 209  |         let source_text = self.allocator.alloc(source_text.to_string());

We're allocating a `String` into arena, so it doesn't get dropped = memory leak. Should allocate a `&str` instead with `Allocator::alloc_str`.

Boshen committed:
feat(syntax_operations): add crate `oxc_syntax_operations` (#6202)

I intend to add constant folding and eval logic to this crate. There are downstream tools that require these functionalities alone.

It's also reasonable to move these traits out of the `ast` crate.
